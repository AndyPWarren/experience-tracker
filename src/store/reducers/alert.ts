import { ActionTypes, IAlertPayload, IAlertDialogue, IClearDialogue } from './../actions';

type Actions = 
	IAlertDialogue |
	IClearDialogue;

export function alert(state: IAlertPayload | null = null, action: Actions ): IAlertPayload | null {
	switch (action.type) {
		case ActionTypes.UpdateAlert:
			return action.payload;

		case ActionTypes.ClearAlert:
			return null;

		default:
			return state;
	}
}
