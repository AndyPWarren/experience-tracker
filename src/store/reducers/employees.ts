import { ActionTypes, IAddEmployee, IUpdateEmployee, IDeleteEmployee } from "../actions";

export interface ICompetence {
	id: number;
	title: string;
	yearsExperience: number;
}

export interface IEmployee {
	id: number;
	name: string;
	totalYearsExperience: number;
	competencies: ICompetence[];
}

export type Action =
	| IAddEmployee
	| IUpdateEmployee
	| IDeleteEmployee;
/**
 * employees reducer
 * @param state array of pervious employee
 * @param action action to perform on the state
 */
export function employees(state: IEmployee[] = [], action: Action): IEmployee[] {
	switch (action.type) {
		case ActionTypes.AddEmployee:
			return [...state, {
				id: (state.length > 0) ? state[state.length - 1].id + 1 : 1,
				name: action.payload.name,
				totalYearsExperience: action.payload.totalYearsExperience,
				competencies: [...action.payload.competencies]
			}];
		case ActionTypes.UpdateEmployee:
			return state.map((employee) => {
				if (employee.id !== action.payload.id) {
					return employee;
				}

				return {
					...employee,
					name: action.payload.name,
					totalYearsExperience: action.payload.totalYearsExperience,
					competencies: [...action.payload.competencies]
				};
			});
		case ActionTypes.DeleteEmployee:
			return state.filter((employee) => employee.id !== action.payload.id);
		default:
			return state;
	}
}
