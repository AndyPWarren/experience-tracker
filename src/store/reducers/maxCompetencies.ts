import { ActionTypes, IUpdateMaxSimCompetencies } from './../actions';
/**
 * max simultaneous competencies reducer
 * @param state previous max simultaneous competencies number
 * @param action action to perform on the state
 */
export function maxSimultaneousCompetencies(state: number = 3, action: IUpdateMaxSimCompetencies ): number {
	switch (action.type) {
		case ActionTypes.UpdateMaxSimCompetencies:
			return action.payload.value;
		default:
			return state;
	}
}
