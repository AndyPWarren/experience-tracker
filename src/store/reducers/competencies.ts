import { Action } from "./employees";
import { ActionTypes } from "../actions";

export interface ICompetence {
	id: number;
	title: string;
	yearsExperience: number;
}

export function competencyReducer(competencies: ICompetence[] = [], action: Action): ICompetence[] {
	switch (action.type) {
		case ActionTypes.AddCompetence:
			return [
				...competencies,
				{
					id: (competencies.length > 0) ? competencies[competencies.length - 1].id + 1 : 1,
					title: action.payload.title,
					yearsExperience: action.payload.yearsExperience
				}
			];
		case ActionTypes.DeleteCompetence:
			return competencies.filter((competence) => competence.id !== action.payload.id);
		case ActionTypes.UpdateCompetence:
			return competencies.map((competency) => {
				if (competency.id !== action.payload.id) {
					return competency;
				}
				return {
					...competency,
					yearsExperience: action.payload.yearsExperience
				};
			});
		default:
			return competencies;
	}
}
