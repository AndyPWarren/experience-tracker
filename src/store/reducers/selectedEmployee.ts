import { ISelectEmployee, ActionTypes } from './../actions';
import { IEmployees, IEmployee } from './employees';

export function selectedEmployeeReducer(state: IEmployees, action: ISelectEmployee ): IEmployee | null {
	switch (action.type) {
		case ActionTypes.SelectEmployee:
			return state.allEmployees.filter((employee) => employee.id === action.payload.id)[0];
		default:
			return null;
	}
}
