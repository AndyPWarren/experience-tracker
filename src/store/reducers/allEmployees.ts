import { IEmployee, Action } from "./employees";
import { ActionTypes } from "../actions";
import { competencyReducer } from "./competencies";

export function allEmployeesReducer(allEmployees: IEmployee[] = [], action: Action): IEmployee[] {
	switch (action.type) {
		case ActionTypes.AddEmployee:
			return [...allEmployees, {
				id: (allEmployees.length > 0) ? allEmployees[allEmployees.length - 1].id + 1 : 1,
				name: action.payload.name,
				totalYearsExperience: action.payload.totalYearsExperience,
				competencies: []
			}];
		case ActionTypes.UpdateEmployee:
			return allEmployees.map((employee) => {
				if (employee.id !== action.payload.id) {
					return employee;
				}

				return {
					...employee,
					name: action.payload.name,
					totalYearsExperience: action.payload.totalYearsExperience
				};
			});
		case ActionTypes.AddCompetence: 
			return allEmployees.map((employee) => {
				if (employee.id !== action.payload.userId) {
					return employee;
				}
				
				return {
					...employee,
					competencies: competencyReducer(employee.competencies, action)
				};
			});
		case ActionTypes.DeleteCompetence: 
			return allEmployees.map((employee) => {
				if (employee.id !== action.payload.userId) {
					return employee;
				}
				
				return {
					...employee,
					competencies: competencyReducer(employee.competencies, action)
				};
			});
		case ActionTypes.UpdateCompetence:
			return allEmployees.map((employee) => {
				if (employee.id !== action.payload.userId) {
					return employee;
				}
				
				return {
					...employee,
					competencies: competencyReducer(employee.competencies, action)
				};
			});
		default:
			return allEmployees;
	}
}
