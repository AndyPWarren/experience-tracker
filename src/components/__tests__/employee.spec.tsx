import { IEmployee } from '../../store/reducers/employees';
import * as renderer from 'react-test-renderer';
import * as React from 'react';
import Employee from '../employee';
import { MemoryRouter } from 'react-router-dom';

describe('Employee Component', () => {
	let mockEmployee: IEmployee;

	beforeEach(() => {
		mockEmployee = {
			id: 1,
			name: 'dave smith',
			totalYearsExperience: 1,
			competencies: []
		};
	});

	it('renders correctly', () => {
		const tree = renderer
			.create(
				<MemoryRouter>
					<Employee employee={mockEmployee} deleteClickHandler={() => console.log} />
				</MemoryRouter>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

