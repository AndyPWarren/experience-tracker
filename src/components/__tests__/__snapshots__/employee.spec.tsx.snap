// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employee Component renders correctly 1`] = `
<div>
  <div
    className="MuiPaper-root-7 MuiPaper-elevation2-11 MuiPaper-rounded-8 MuiCard-root-6 EmployeeCard-card-1"
  >
    <div
      className="MuiCardHeader-root-34"
    >
      <div
        className="MuiCardHeader-avatar-35"
      >
        <div
          aria-label="Recipe"
          className="MuiAvatar-root-40 MuiAvatar-colorDefault-41 EmployeeCard-avatar-5"
        >
          d
        </div>
      </div>
      <div
        className="MuiCardHeader-content-37"
      >
        <span
          className="MuiTypography-root-43 MuiTypography-body2-51 MuiCardHeader-title-38"
        >
          dave smith
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root-67"
    >
      <p
        className="MuiTypography-root-43 MuiTypography-body1-52"
      >
        Total years of experience: 
        1
      </p>
    </div>
    <div
      className="MuiCardActions-root-68 EmployeeCard-actions-2"
    >
      <a
        href="/edit-employee/1"
        onClick={[Function]}
      >
        <button
          aria-label="Edit"
          className="MuiButtonBase-root-76 MuiIconButton-root-70"
          disabled={false}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiIconButton-label-75"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-79"
              color={undefined}
              focusable="false"
              viewBox="0 0 24 24"
            >
              <g>
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </g>
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-85"
          />
        </button>
      </a>
      <button
        aria-label="Edit"
        className="MuiButtonBase-root-76 MuiIconButton-root-70"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-75"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-79"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <g>
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </g>
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-85"
        />
      </button>
    </div>
  </div>
</div>
`;
